coderd/notifications.go

```diff
@@ -323,6 +342,91 @@ func (api *API) putUserNotificationPreferences(rw http.ResponseWriter, r *http.R
 	httpapi.Write(ctx, rw, http.StatusOK, out)
 }
 
+// @Summary Send a custom notification
+// @ID send-a-custom-notification
+// @Security CoderSessionToken
+// @Tags Notifications
+// @Accept json
+// @Produce json
+// @Param request body codersdk.CustomNotificationRequest true "Provide a non-empty title or message"
+// @Success 204 "No Content"
+// @Failure 400 {object} codersdk.Response "Invalid request body"
+// @Failure 403 {object} codersdk.Response "System users cannot send custom notifications"
+// @Failure 500 {object} codersdk.Response "Failed to send custom notification"
+// @Router /notifications/custom [post]
+func (api *API) postCustomNotification(rw http.ResponseWriter, r *http.Request) {
+	var (
+		ctx    = r.Context()
+		apiKey = httpmw.APIKey(r)
+	)
+
+	// Parse request
+	var req codersdk.CustomNotificationRequest
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	// Validate request: require `content` and at least one non-empty `title` or `message`.
+	if err := req.Validate(); err != nil {
+		api.Logger.Error(ctx, "send custom notification: validation failed", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid request body",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	// Block system users from sending custom notifications
+	user, err := api.Database.GetUserByID(ctx, apiKey.UserID)
+	if err != nil {
+		api.Logger.Error(ctx, "send custom notification", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to send custom notification",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	if user.IsSystemUser() {
+		api.Logger.Error(ctx, "send custom notification: system user is not allowed",
+			slog.F("id", user.ID.String()), slog.F("name", user.Name))
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Forbidden",
+			Detail:  "System users cannot send custom notifications.",
+		})
+		return
+	}
+
+	if _, err := api.NotificationsEnqueuer.EnqueueWithData(
+		//nolint:gocritic // We need to be notifier to send the notification.
+		dbauthz.AsNotifier(ctx),
+		user.ID,
+		notifications.TemplateCustomNotification,
+		map[string]string{
+			"custom_title":   req.Content.Title,
+			"custom_message": req.Content.Message,
+		},
+		map[string]any{
+			// Current dedupe is done via an hash of (template, user, method, payload, targets, day).
+			// We intentionally include a timestamp to bypass the per-day dedupe so the caller can
+			// resend identical content to themselves multiple times in one day.
+			// TODO(ssncferreira): When we support sending custom notifications to multiple users/roles,
+			//   enforce proper deduplication across recipients to reduce noise and prevent spam.
+			//   See https://github.com/coder/coder/issues/19768
+			"dedupe_bypass_ts": api.Clock.Now().UTC(),
```

Comment:
suggest rounding this to `time.Minute` so users can't spam themselves every millisecond